abel/lib/direct.pyx

```diff
@@ -46,48 +46,38 @@ cpdef _cabel_direct_integral(double [:, ::1] f, double [::1] r, int correction):
 
     with nogil:
         # pre-calculate the array of r_i, r_j
-        for i in range(N1):
-            for j in range(0,i):
-                val = sqrt(r[i]**2 - r[j]**2)
-                I_sqrt[j,i] = val
-                I_isqrt[j,i] = 1./val
+        for j in range(N1):
+            for k in range(j+1, N1):
+                val = sqrt(r[k]**2 - r[j]**2)
+                I_sqrt[j,k] = val
+                I_isqrt[j,k] = 1./val
 
         for i in range(N0): # loop over rows (z)
-            for j in range(N1 - 1):  # loop over (r) elements
-                # Integrating with the Simpson rule the part of the
-                # integral that is numerically stable
-                # https://en.wikipedia.org/wiki/Simpson%27s_rule#Python
-                # We use this for both odd and even cases, while the
-                # Simpson rule is only applicable in the even case.
-                # However, we still use this wrong version as it has
-                # demonstrated good results empirically.
-                s = f[i,j+1]*I_isqrt[j,j+1] + f[i,N1-1]*I_isqrt[j,N1-1]
-                for k in range(j+2, N1-1): # inner loop over elements such as r < y 
-                    val = f[i,k]*I_isqrt[j,k]
-                    if (k-j+1) % 2 == 0:
-                        s = s + 2*val
-                    else:
-                        s = s + 4*val
-                s = s * dr / 3.
-
+            for j in range(N1):  # loop over (r) elements
+        
+                s = 0
+            
+                # simple left Reimann sum:
+                # for k in range(j+1, N1): # inner loop over elements such as r < y
+                #   s = s + f[i,k] * I_isqrt[j,k] * dr
+            
+                # Trapezoidal rule integration, skipping r == y
+                for k in range(j+1, N1-1): # inner loop over elements such as r < y
+                    s = s + f[i,k] * I_isqrt[j,k] * 0.5 * dr  + f[i,k+1] * I_isqrt[j,k+1] * 0.5 * dr
```

Comment:
maybe use inplace add and factorize multiplications?
```py
s += (f[..] * I_isqrt[..] + f[..] * I_isqrt[..]) * 0.5 * dr
```

Reply:
Very clever!

Reply:
Sounds good to me!